nxcomp/src/Timestamp.cpp

```diff
@@ -44,34 +44,55 @@
 
 T_timestamp timestamp;
 
-//
-// The following functions all use the ctime
-// static buffer from the C library.
-//
-
-char *strTimestamp(const T_timestamp &ts)
+std::string strTimestamp(const T_timestamp &ts)
 {
-  char *ctime_now = ctime((time_t *) &ts.tv_sec);
+  std::string ret;
+
+  char ctime_now[26] = { };
+  bool err = true;
+
+#if HAVE_CTIME_S
+  errno_t retval = ::ctime_s(ctime_now, 26, static_cast<const time_t*>(&ts.tv_sec));
+
+  err = (retval != 0);
+#else
+  char *retval = ::ctime_r(static_cast<const time_t*>(&ts.tv_sec), ctime_now);
 
+  err = (!(retval));
+#endif
+
+  if (err)
+  {
+    std::cerr << "WARNING: converting time to string failed." << std::endl;
+  }
+
+  /* Replace newline at position 25 with a NULL byte. */
   ctime_now[24] = '\0';
 
-  return ctime_now;
+  ret = ctime_now;
+
+  return ret;
 }
 
 //
-// This is especially dirty. 
+// This is especially dirty.
 //
 
-char *strMsTimestamp(const T_timestamp &ts)
+std::string strMsTimestamp(const T_timestamp &ts)
 {
-  char *ctime_now = ctime((time_t *) &ts.tv_sec);
+  std::string ret;
+
+  std::string ctime_now = strTimestamp(ts);
 
-  char ctime_new[25];
+  if (!(ctime_now.empty()))
+  {
+    char ctime_new[26] = { };
 
-  sprintf(ctime_new, "%.8s:%3.3f", ctime_now + 11,
-              (float) ts.tv_usec / 1000);
+    snprintf(ctime_new, sizeof(ctime_new), "%.8s:%3.3f",
+             ctime_now.c_str() + 11, static_cast<float>(ts.tv_usec) / 1000);
 
-  strncpy(ctime_now, ctime_new, 24);
+    ret = ctime_new;
+  }
 
-  return ctime_now;
+  return ret;
```

Comment:
ret is uninitilaized here if ctime_now is empty!

Reply:
`std::string` is a C++ class and we can thus trust it to initialize its internal data in a sane way. This statement will use the default constructor, which creates/initializes an empty `std::string`.

Way different from C behavior.

Reply:
ok